### PR TITLE
refactor(limits): Prefer to use single API for commits, PR and branches

### PR DESCRIPTION
**Pull request by zharinov**
Ref: #7076

This PR is meant to be non-destructive

### CODE REVIEW

File: lib/workers/repository/process/write.ts [hunk 1]
```diff
@@ -27,31 +27,28 @@ export async function writeUpdates(
...
+      incrementLimit('prsRemaining');
     }
     // istanbul ignore if
     if (
       res === 'automerged' &&
       branch.automergeType === 'pr-comment' &&
       branch.requiredStatusChecks === null
     ) {
-      deductPrRemainingCount = 1;
+      incrementLimit('prsRemaining');
```
Comment (zharinov) [inline lib/workers/repository/process/write.ts:50]: In more aggressive refactoring, we can move this `incrementLimit` calls right after the `platform.createPr(...)` calls we're interested in. In fact, I would do it here, but not sure about this `automerge` branch).

**Comment by zharinov** [comment]
Extracted `limits.ts` refactoring to separate PR: #7140

**Comment by zharinov** [comment]
Waiting for #7140

**Comment by zharinov** [comment]
Should not be destructive, but I'll make test run anyway

**Comment by zharinov** [comment]
Can confirm it worked for both commit limits and PR limits

File: lib/workers/global/limits.ts [hunk 2]
```diff
@@ -15,15 +16,17 @@ export function resetAllLimits(): void {
...
+  if (typeof max === 'number') {
+    const maxVal = max > 0 || (allowZero && max <= 0) ? Math.max(max, 0) : null;
+    logger.debug(`${key} limit = ${maxVal}`);
+    const limit = limits.get(key);
+    limits.set(key, {
+      current: 0,
+      ...limit,
+      max: maxVal,
+    });
+  }
```
Comment (viceice) [inline lib/workers/global/limits.ts:29]: Shouldn't this change be a seperate pr?

**Review by viceice — COMMENTED** [review]
just a small thing, otherwise LGTM

File: lib/workers/global/limits.ts (hunk 2 above)
Comment (zharinov) [reply lib/workers/global/limits.ts:29]: Maybe, just thought it's not so substantial and we already have too much preparatory PRs for branch concurrent limits

**Comment by rarkins** [comment]
How thoroughly tested is this?

I feel a little uncomfortable about how some of the limit logic is still out in the worker layer, e.g. we have three types of limits in total:
- per run commits
- per repo PRs
- per hour PRs

But within the limits they're using the same data structure.

**Comment by zharinov** [comment]
As far as we are processing repositories sequentially, these change shouldn't alter current behaviour. It basically changes how we store and retrieve limit information. i.e. via `workers/global/limits.ts` module instead of stack-local boolean flags.

But the mechanism of how we compute those limits wasn't changed: we're still basically using `Math.min(getPrHourlyRemaining(...), getConcurrentPrsRemaining(...)`, which is computed per-repo and right before branches processing. And as well, we're setting commit limitations once we parsed CLI config.

As the next step, I planned to introduce also `Limit.Branch` element, i.e. this `Limit` thing meant to provide keys for the entities that Renovate can create, but user can limit. As we introduce new config options, we may change the way how those limits are calculated (and how frequently).

**Comment by zharinov** [comment]
I can confirm all limits worked for me in different combinations. By the way, for `prHourlyLimit` logger can show negative remaining PRs (I closed too much), but for the next hour it worked fine. I'll fix this number later, here I'm intentionally not touching limit calculation code.

As an idea, we can create additional tracking of which particular limit was fired, for better logging.

**Review by viceice — DISMISSED** [review]

**Comment by zharinov** [comment]
Sorry, I think the recent commit fixes the very wrong thing

File: lib/workers/global/limits.ts [hunk 3]
```diff
@@ -15,15 +16,15 @@ export function resetAllLimits(): void {
   limits.clear();
 }
 
-export function setMaxLimit(key: Limit, max: unknown): void {
-  const maxVal = typeof max === 'number' && max > 0 ? max : null;
-  logger.debug(`${key} limit = ${maxVal}`);
-  const limit = limits.get(key);
-  limits.set(key, {
-    current: 0,
-    ...limit,
```
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/workers/global/limits.ts:24]: Right, and I think its calls are located in right places

**Review by viceice — APPROVED** [review]

**Comment by zharinov** [comment]
Superseded by #7198

**Pull request closed without merging**